src/include/platform/internal/GenericConfigurationManagerImpl.h

```diff
@@ -77,6 +77,13 @@ class GenericConfigurationManagerImpl
     CHIP_ERROR _GetDeviceCertificate(uint8_t * buf, size_t bufSize, size_t & certLen);
     CHIP_ERROR _GetDeviceIntermediateCACerts(uint8_t * buf, size_t bufSize, size_t & certsLen);
     CHIP_ERROR _GetDevicePrivateKey(uint8_t * buf, size_t bufSize, size_t & keyLen);
+    CHIP_ERROR _GetDeviceDACCertificate(uint8_t * buf, size_t bufSize, size_t & dacLen);
+    CHIP_ERROR _GetDevicePAICertificate(uint8_t * buf, size_t bufSize, size_t & paiLen);
+    CHIP_ERROR _GetDevicePAACertificate(const uint8_t * skid, size_t skidLen, uint8_t * buf, size_t bufSize, size_t & paaLen);
+    CHIP_ERROR _GetDeviceCertificationDeclaration(uint8_t * buf, size_t bufSize, size_t & certDeclLen);
+    CHIP_ERROR _GetDeviceAttestationSignature(const uint8_t * attestationElements, size_t attestationElementsSize,
```

Comment:
The attestation primitive must be more generic and not rely on attestation elements since it's *also required for CSR*.

I have a suggested interface:
```
namespace chip {

class DeviceAttestationCredentials
{
public:
    DeviceAttestationCredentials() = default;
    virtual ~DeviceAttestationCredentials() = default;

    // Not copyable
    DeviceAttestationCredentials(const DeviceAttestationCredentials&) = delete;
    DeviceAttestationCredentials& operator=(const DeviceAttestationCredentials&) = delete

    virtual CHIP_ERROR GetCertificationDeclaration(MutableByteSpan & out_cd_buffer) = 0;
    virtual CHIP_ERROR GetFirmwareInformation(MutableByteSpan & out_firmware_info_buffer) = 0;
    virtual CHIP_ERROR GetDeviceAttestationCert(MutableByteSpan & out_dac_buffer) = 0;
    virtual CHIP_ERROR GetProductAttestationIntermediateCert(MutableByteSpan & out_pai_buffer) = 0;
    virtual CHIP_ERROR SignWithDeviceAttestationKey(const ByteSpan & payload, MutableByteSpan & out_signature_buffer) = 0;
};

}  // namespace chip

```

The above interface could be implemented per platform, and supports:

* Accessing the necessary fields for a device being attested, wtihout needing access to PAA
* Signing with the device attestation key without relying on having direct access to the key, and without relying on particular format

For the PAA certificates, I would argue that:

1. Device attestation signature + chain verification should be abstracted in a delegate that different platforms can implement.
2. While an example for chip-tool's commissioning code can be employed, it's important not to bake-in the assumption of direct access to PAA possible. It's already clear in spec that direct access to all PAAs is not mandatory at the commissioner itself (although it's feasible), due to the variety of device attestation policies.

For example: an ecosystem with full end-to-end attestation validation may require verifying the device attestation signature of the CSRResponse **at the CA** to ensure that the CA is not used as a signing oracle for operational credentials. Because of this, we should not bake a 1-size-fits all *validation* path with mandatory access to PAAs.

Reply:
@tcarmelveilleux  are you suggesting to define a separate class instead of using CRTP pattern as it is currently implemented in this file?

Reply:
@emargolis  Yes, just keep away from ConfigurationManager altogether.